### PR TITLE
Add option for Empty Attributes

### LINKED ISSUE #128: Boolean attributes quotes pattern
## Request / Idea
Give the possibility to choose between use empty quotes or remove them when working with boolean attributes.

## Input

```pug
input(readonly="" disabled)
```

### `'none'`
```pug
input(readonly disabled)
```

### `'all'`
```pug
input(readonly="" disabled="")
```

### `'as-is'` **default**
```pug
input(readonly="" disabled)
```

### `'all'` with _exceptions_:
```javascript
const pugEmptyAttributes = 'all'
const pugEmptyAttributesExceptions = ['^@']
```

```pug
input(readonly disabled @click="")
```

### ~_as-needed_ ? on Vue or Angular this also may be necessary~
([why not](https://github.com/prettier/plugin-pug/issues/128#issuecomment-704034813))

```pug
input(readonly disabled @click="")
```

### PR DESCRIPTION
**Pull request by SkyaTura**
closes #128 

### Todo

- [x] Add tests
- [x] Add option descriptions
- [x] Add documentation
- [x] Use native prettier typedefs

### CODE REVIEW

**Comment by Shinigami92** [comment]
I have update the deps in master, pull it :)

File: src/options/empty-attributes.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
+import { AttributeToken } from 'pug-lexer';
+import { ChoiceSupportOption, PathArraySupportOption } from 'prettier';
+import { CATEGORY_PUG } from '.';
+
+// eslint-disable-next-line @typescript-eslint/typedef
+export const PUG_EMPTY_ATTRIBUTES_OPTION: ChoiceSupportOption = {
+	since: '1.9.0',
```
Comment (Shinigami92) [inline src/options/empty-attributes.ts:7]: 1.9.0 is already released, I think the next one would be 1.10.0

File: src/options/empty-attributes.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		}
+	]
+};
+
+// eslint-disable-next-line @typescript-eslint/typedef
+export const PUG_EMPTY_ATTRIBUTES_FORCE_QUOTES_OPTION: PathArraySupportOption = {
+	since: '1.9.0',
+	category: CATEGORY_PUG,
+	type: 'path',
+	default: ([{ value: [] }] as unknown) as string[],
```
Comment (Shinigami92) [inline src/options/empty-attributes.ts]:
```suggestion
	default: [{ value: [] }],
```

File: src/options/empty-attributes.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
+import { AttributeToken } from 'pug-lexer';
+import { ChoiceSupportOption, PathArraySupportOption } from 'prettier';
+import { CATEGORY_PUG } from '.';
+
+// eslint-disable-next-line @typescript-eslint/typedef
```
Comment (Shinigami92) [inline src/options/empty-attributes.ts:5]: You now can remove all these `disable` statements
We have updated prettier types

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -242,6 +242,21 @@ These additional options are specific to pug templates and can be configured in
...
   _default_: `false`
 
+- `pugEmptyAttributes`  
+  Change behavior of boolean attributes.
+
+  Choices:
+
+  - `'as-is'` _default_ -> Nothing is changed.  
+    Example: `foo(a, b="", c)`
+  - `'none'` -> Every attribute with empty quotes will be replaced by a boolean version without quotes.
```
Comment (Shinigami92) [inline README.md:252]: I'm not quite satisfied with the `replaced by a boolean version` part :thinking:  
It's a little bit confusing

**Review by Shinigami92 — CHANGES_REQUESTED** [review]

**Comment by Shinigami92** [comment]
Seems CI is stuck, will come back to it tomorrow

File: README.md (hunk 4 above)
Comment (SkyaTura) [reply README.md:252]: What about "Every attribute with empty quotes will have them removed"

Comment (Shinigami92) [reply README.md:252]: At least, a bit better

Applies also for https://github.com/prettier/plugin-pug/pull/142/files#diff-dd2c9e3ce59123522b95af9ce90edd8bceac03275e49d64680605524c75fd126R18

**Comment by Shinigami92** [comment]
... GitHub is a bit buggy currently :thinking: 
Will wait after CI reaches 24h, after that I will contact the support...

**Comment by Shinigami92** [comment]
I have contacted the GitHub support

**Comment by SkyaTura** [comment]
@Shinigami92 why is the test failing now?

**Comment by Shinigami92** [comment]
I'm not quite sure but one of the test is failing with following message:



Can you reproduce this locally?

Maybe also have a look over my changes?
I'm not totally sure but it may come from updating to `@types/node": "~14.14.0`  

https://github.com/prettier/plugin-pug/blob/a2a87af9980f8acbbc27b105539d33e7ee7aeb79/package.json#L40

IMO this doesn't make sense, but I don't know it better 🤔 

I'm confused why all other tests are working

**Comment by Shinigami92** [comment]
Okay... it's NOT `"@types/node": "~14.14.0"`

I have rolled back some `devDependencies` locally and this is not the source of the problem :thinking: 

I'm currently totally confused...

What does `Unexpected type undefined` mean?

It has something to do with this: https://github.com/prettier/prettier/blob/2.1.2/src/main/options-normalizer.js#L122-L166

And therefor potentially something with:

https://github.com/prettier/plugin-pug/blob/9b96875918d4cf00c6ef527dfd89f5cce997cc2d/src/options/empty-attributes.ts#L8
and/or
https://github.com/prettier/plugin-pug/blob/9b96875918d4cf00c6ef527dfd89f5cce997cc2d/src/options/empty-attributes.ts#L30

But why did your commit `Update release version` pass? That's so confusing 😕

**Comment by Shinigami92** [comment]
WHAT. THE. FUCK. :eyes:

**Comment by SkyaTura** [comment]
Now that you've mentioned, I noticed your "organize imports" change.

I've struggled with this problem for a while back in the beginning of the tests implementation, and I have not a mostly remote idea of why on earth the plugin must be imported first... but it does.

I can't understand what have I done wrong, since everything else works just fine.

:P

**Comment by Shinigami92** [comment]
I need to investigate if this is a bug in jest or ts-jest :/

Maybe I will merge this PR in the next few days

**Review by Shinigami92 — APPROVED** [review]

**Comment by SkyaTura** [comment]
That's odd, but thank you anyway ;)

**Pull request merged**